server/src/main/java/io/crate/expression/eval/EvaluatingNormalizer.java

```diff
@@ -238,6 +239,12 @@ private Symbol normalizeFunction(Function function, TransactionContext context)
                 context.sessionSettings().searchPath()
             );
             assert implementation != null : "Function implementation not found using full qualified lookup: " + function;
+
+            // Some scalar implementation may change their semantic by compile as session settings are passed it.
+            if (implementation instanceof Scalar scalar) {
+                implementation = scalar.compile(function.arguments(), context.sessionSettings(), nodeCtx.userLookup());
```

Comment:
Calling `compile` on normalization currently causes some tests to fail as some scalars do not take care of `NULL` literals inside their `compile` methods.
This would be required if we want to adjust the return type of the cast functions once by creating a new instance.
The only drawback I see is that it may not conceptually fit and successful normalization results in a literal so compilation in order to avoid repeated expensive logic on every row won't count.

Another option would be to adjust the return type additionally inside the casts `normalize` method.
The drawback here would be a bit of duplicated code.

@mfussenegger wdyt?

Reply:
I'm thinking that `evaluate` should probably have this additional logic as well. `compile` should be a pure performance optimization with no semantics change.

Reply:
> I'm thinking that evaluate should probably have this additional logic as well. 

In this case we'd have to
a) modify the used type for casting on every row (which has probably performance impacts)
b) mutate the `returnType` of the instance.

Reply:
Isn't it possible to still have the `compile` to avoid the per row performance implications, but have `evaluate` / `normalizeSymbol` do the check too?

Or given that the `returnType` is available at construction time, we could have two different cast implementations without using compile for it?

Reply:
> Isn't it possible to still have the compile to avoid the per row performance implications, but have evaluate / normalizeSymbol do the check too?

Doing it on `compile` + `normalizeSymbol` would work, but why then also on `evaluate` (as evaluation is called per row)? 

> Or given that the returnType is available at construction time, we could have two different cast implementations without using compile for it?

Maybe I misunderstand your point here, but that is exactly the root cause. The `search_path` of the session is currently not passed-in to the constructors on function resolving. Do you argue that this should be changed (which could be a solution indeed)?

Reply:
> Doing it on compile + normalizeSymbol would work, but why then also on evaluate (as evaluation is called per row)?

Because both `compile` and `normalizeSymbol` are so far kinda optional. I think `SymbolEvaluator` for example doesn't call them.

> Maybe I misunderstand your point here, but that is exactly the root cause. The search_path of the session is currently not passed-in to the constructors on function resolving. Do you argue that this should be changed (which could be a solution indeed)?

It wasn't what I had in mind but I was confused by the second constructor that's called from compile. Maybe it would be an option. Or maybe as alternative we specialize the regclass cast case completely without going through the generic cast? Not sure if that works better or is worse.

I also just saw that this adds the `currentSchema` to the `Regclass` type, but doesn't stream the value (so data type serialization would also change semantics). I think we've to address that too somehow.